File: kerf/parametric/expr.py
"""Parameter expressions.

Any numeric field in a part file may be written as an expression over the
parameter table, so `"radius": "bolt_d/2"` stays correct when bolt_d moves.
Expressions are read by walking a parsed syntax tree. Python's eval is never
called, because a part file is data that arrives from other people.
"""

from __future__ import annotations

import ast
import math
from typing import Any

import numpy as np

ALLOWED_FUNCTIONS = {
    "sqrt": math.sqrt, "sin": math.sin, "cos": math.cos, "tan": math.tan,
    "abs": abs, "min": min, "max": max, "round": round, "floor": math.floor,
    "ceil": math.ceil, "atan2": math.atan2, "radians": math.radians,
    "degrees": math.degrees, "hypot": math.hypot, "pow": math.pow,
}

ALLOWED_CONSTANTS = {"pi": math.pi, "tau": math.tau, "e": math.e}

# An exponent large enough to make Python build a number with millions of
# digits, which is a way to hang a rebuild with nothing but a part file.
MAX_EXPONENT = 1024.0


class ExpressionError(ValueError):
    """Raised when an expression is malformed or uses something not allowed."""


def _power(base: float, exponent: float) -> float:
    """Raise to a power, refusing exponents that would take a visible time.

    A part file arrives from somebody else, and `2**(2**30)` is a plain
    arithmetic expression that Python will happily spend minutes on.
    """
    if abs(exponent) > MAX_EXPONENT:
        raise ExpressionError(f"exponent {exponent:g} is out of range")
    return base ** exponent


BINARY_OPERATORS = {
    ast.Add: lambda a, b: a + b,
    ast.Sub: lambda a, b: a - b,
    ast.Mult: lambda a, b: a * b,
    ast.Div: lambda a, b: a / b,
    ast.Pow: _power,
    ast.Mod: lambda a, b: a % b,
    ast.FloorDiv: lambda a, b: a // b,
}


class ExpressionError(ValueError):
    """Raised when an expression is malformed or uses something not allowed."""



def evaluate_expression(expression: str, params: dict[str, float]) -> float:
    """Evaluate one expression against a parameter table."""
    try:
        tree = ast.parse(expression, mode="eval")
    except SyntaxError as error:
        raise ExpressionError(f"cannot parse expression {expression!r}: {error}") from error

    def walk(node: ast.AST) -> float:
        if isinstance(node, ast.Expression):
            return walk(node.body)
        if isinstance(node, ast.Constant):
            if isinstance(node.value, (int, float)) and not isinstance(node.value, bool):
                return float(node.value)
            raise ExpressionError(f"unsupported constant {node.value!r}")
        if isinstance(node, ast.Name):
            if node.id in params:
                return float(params[node.id])
            if node.id in ALLOWED_CONSTANTS:
                return ALLOWED_CONSTANTS[node.id]
            raise ExpressionError(f"unknown parameter {node.id!r}")
        if isinstance(node, ast.BinOp) and type(node.op) in BINARY_OPERATORS:
            left, right = walk(node.left), walk(node.right)
            return _arithmetic(
                BINARY_OPERATORS[type(node.op)], left, right, expression=expression
            )
        if isinstance(node, ast.UnaryOp):
            if isinstance(node.op, ast.USub):
                return -walk(node.operand)
            if isinstance(node.op, ast.UAdd):
                return walk(node.operand)
        if isinstance(node, ast.Call) and isinstance(node.func, ast.Name):
            function = ALLOWED_FUNCTIONS.get(node.func.id)
            if function is None:
                raise ExpressionError(f"function {node.func.id!r} is not allowed")
            arguments = [walk(argument) for argument in node.args]
            return _arithmetic(function, *arguments, expression=expression)
        raise ExpressionError(f"unsupported syntax in expression {expression!r}")

    value = walk(tree)
    if not math.isfinite(value):
        raise ExpressionError(f"expression {expression!r} does not resolve to a finite number")
    return value


def _arithmetic(operation, *arguments, expression: str) -> float:
    """Apply one operation, turning every arithmetic failure into our own error.

    Division by zero, the square root of a negative number, and an overflow
    are all things a parameter table can ask for. Each raises its own builtin
    exception, and letting those escape means a rebuild error arrives as a
    traceback instead of as a message naming the equation.
    """
    try:
        return float(operation(*arguments))
    except ExpressionError:
        raise
    except ZeroDivisionError as error:
        raise ExpressionError(f"{expression!r} divides by zero") from error
    except (ArithmeticError, ValueError, TypeError) as error:
        raise ExpressionError(f"cannot evaluate {expression!r}: {error}") from error


def resolve(value: Any, params: dict[str, float]) -> float:
    """Turn a literal or an expression into a number."""
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return float(value)
    if isinstance(value, str):
        return evaluate_expression(value, params)
    raise ExpressionError(f"expected a number or expression, got {value!r}")


def resolve_vec(value: Any, params: dict[str, float], default=(0.0, 0.0, 0.0)) -> np.ndarray:
    """Turn a three element list of literals or expressions into a vector."""
    if value is None:
        return np.asarray(default, dtype=float)
    return np.asarray([resolve(item, params) for item in value], dtype=float)


def expression_dependencies(value: Any) -> set[str]:
    """Names an expression reads.

    This is what lets kerf answer "what does this parameter drive". When
    bolt_d changes, the features whose expressions mention bolt_d are the
    features that changed with it.
    """
    if not isinstance(value, str):
        return set()
    try:
        tree = ast.parse(value, mode="eval")
    except SyntaxError:
        return set()
    return {
        node.id
        for node in ast.walk(tree)
        if isinstance(node, ast.Name)
        and node.id not in ALLOWED_CONSTANTS
        and node.id not in ALLOWED_FUNCTIONS
    }

File: web/engine/expr.js
// Parameter expressions.
//
// A part file may write any number as an expression over the parameter table,
// so "bolt_d/2" stays correct when bolt_d moves. The expression is parsed into
// a small tree and walked. eval is never used, because a part file is data
// that arrives from other people.

const FUNCTIONS = {
  sqrt: Math.sqrt, sin: Math.sin, cos: Math.cos, tan: Math.tan,
  abs: Math.abs, min: Math.min, max: Math.max, round: Math.round,
  floor: Math.floor, ceil: Math.ceil, atan2: Math.atan2, hypot: Math.hypot,
  pow: Math.pow,
  radians: (d) => (d * Math.PI) / 180,
  degrees: (r) => (r * 180) / Math.PI,
};

const CONSTANTS = { pi: Math.PI, tau: Math.PI * 2, e: Math.E };

export class ExpressionError extends Error {}

function tokenize(text) {
  const tokens = [];
  let index = 0;
  while (index < text.length) {
    const char = text[index];
    if (/\s/.test(char)) { index += 1; continue; }
    if (/[0-9.]/.test(char)) {
      let end = index;
      while (end < text.length && /[0-9.eE]/.test(text[end])) {
        // An e only continues the number when a sign or digit follows it.
        if (/[eE]/.test(text[end]) && !/[0-9+-]/.test(text[end + 1] || '')) break;
        if (/[eE]/.test(text[end]) && /[+-]/.test(text[end + 1])) end += 1;
        end += 1;
      }
      const value = Number(text.slice(index, end));
      if (Number.isNaN(value)) throw new ExpressionError(`bad number in ${text}`);
      tokens.push({ type: 'number', value });
      index = end;
      continue;
    }
    if (/[A-Za-z_]/.test(char)) {
      let end = index;
      while (end < text.length && /[A-Za-z0-9_]/.test(text[end])) end += 1;
      tokens.push({ type: 'name', value: text.slice(index, end) });
      index = end;
      continue;
    }
    if ('+-*/%(),'.includes(char)) {
      tokens.push({ type: char });
      index += 1;
      continue;
    }
    if (char === '*' || char === '^') { tokens.push({ type: '^' }); index += 1; continue; }
    throw new ExpressionError(`unexpected character ${char} in ${text}`);
  }
  return tokens;
}

function parse(text) {
  const tokens = tokenize(text.replace(/\*\*/g, '^').replace(/\^/g, '^'));
  let position = 0;

  const peek = () => tokens[position];
  const take = (type) => {
    if (!peek() || peek().type !== type) {
      throw new ExpressionError(`expected ${type} in ${text}`);
    }
    return tokens[position++];
  };

  function primary() {
    const token = peek();
    if (!token) throw new ExpressionError(`unexpected end of ${text}`);
    if (token.type === 'number') { position += 1; return { kind: 'number', value: token.value }; }
    if (token.type === '-') { position += 1; return { kind: 'negate', operand: primary() }; }
    if (token.type === '+') { position += 1; return primary(); }
    if (token.type === '(') {
      position += 1;
      const inner = expression();
      take(')');
      return inner;
    }
    if (token.type === 'name') {
      position += 1;
      if (peek() && peek().type === '(') {
        position += 1;
        const args = [];
        if (peek() && peek().type !== ')') {
          args.push(expression());
          while (peek() && peek().type === ',') { position += 1; args.push(expression()); }
        }
        take(')');
        return { kind: 'call', name: token.value, args };
      }
      return { kind: 'name', name: token.value };
    }
    throw new ExpressionError(`unexpected token in ${text}`);
  }

  function power() {
    const base = primary();
    if (peek() && peek().type === '^') {
      position += 1;
      return { kind: 'binary', op: '^', left: base, right: power() };
    }
    return base;
  }

  function term() {
    let left = power();
    while (peek() && ['*', '/', '%'].includes(peek().type)) {
      const op = tokens[position++].type;
      left = { kind: 'binary', op, left, right: power() };
    }
    return left;
  }

  function expression() {
    let left = term();
    while (peek() && ['+', '-'].includes(peek().type)) {
      const op = tokens[position++].type;
      left = { kind: 'binary', op, left, right: term() };
    }
    return left;
  }

  const tree = expression();
  if (position !== tokens.length) throw new ExpressionError(`trailing input in ${text}`);
  return tree;
}

const cache = new Map();

function parseCached(text) {
  if (!cache.has(text)) cache.set(text, parse(text));
  return cache.get(text);
}

function walk(node, params) {
  switch (node.kind) {
    case 'number': return node.value;
    case 'negate': return -walk(node.operand, params);
    case 'name': {
      if (Object.prototype.hasOwnProperty.call(params, node.name)) return Number(params[node.name]);
      if (node.name in CONSTANTS) return CONSTANTS[node.name];
      throw new ExpressionError(`unknown parameter ${node.name}`);
    }
    case 'call': {
      const fn = FUNCTIONS[node.name];
      if (!fn) throw new ExpressionError(`function ${node.name} is not allowed`);
      return fn(...node.args.map((arg) => walk(arg, params)));
    }
    case 'binary': {
      const a = walk(node.left, params);
      const b = walk(node.right, params);
      if (node.op === '+') return a + b;
      if (node.op === '-') return a - b;
      if (node.op === '*') return a * b;
      if (node.op === '/') return a / b;
      if (node.op === '%') return a % b;
      return a ** b;
    }
    default: throw new ExpressionError('unsupported expression');
  }
}

export function evaluateExpression(text, params) {
  return walk(parseCached(text), params);
}

export function resolve(value, params) {
  if (typeof value === 'number') return value;
  if (typeof value === 'string') return evaluateExpression(value, params);
  throw new ExpressionError(`expected a number or expression, got ${value}`);
}

export function resolveVec(value, params, fallback = [0, 0, 0]) {
  if (value === undefined || value === null) return fallback.slice();
  return value.map((item) => resolve(item, params));
}

// Which parameters an expression reads. This is what lets the diff say that
// changing bolt_d moved four holes.
export function expressionDependencies(value) {
  if (typeof value !== 'string') return new Set();
  let tree;
  try {
    tree = parseCached(value);
  } catch {
    return new Set();
  }
  const found = new Set();
  (function visit(node) {
    if (node.kind === 'name' && !(node.name in CONSTANTS)) found.add(node.name);
    if (node.kind === 'negate') visit(node.operand);
    if (node.kind === 'binary') { visit(node.left); visit(node.right); }
    if (node.kind === 'call') node.args.forEach(visit);
  })(tree);
  return found;
}
